desc/protoparse/linker.go

```diff
@@ -118,6 +118,13 @@ func (l *linker) createDescriptorPool() error {
 				return err
 			}
 		}
+		dependencies := make(map[string]struct{}, len(fd.Dependency))
+		for _, dependency := range fd.Dependency {
+			if _, ok := dependencies[dependency]; ok {
+				return fmt.Errorf("%s: import %q was listed twice", filename, dependency)
```

Comment:
This library generally doesn't use `fmt.Errorf` because the resulting errors do not contain source position information and can be treated as I/O errors or other fatal errors (meaning they can cause compilation to abort prematurely).

Instead of iterating through dependencies in the file descriptor proto, you'd look through the declaration nodes in the file node (root of the AST). For each declaration, if it's a `*ast.ImportNode`, then you can extract the import name and location (`impNode.Name.Start()`) and store the location in the map (instead of `struct{}` as the value type). That way, the error's source position can attribute the later position, and the message can refer to the location of the first import (the value already in the map).

Also, there's no need to delay until link time to check for this. This should go in the basic validation of the descriptor proto instead (see `validateBasic` in `validate.go`).